drv/sidecar-mainboard-controller/src/tofino2.rs

```diff
@@ -231,3 +232,429 @@ impl Sequencer {
+
+        if !state.send_buffer_empty() || !state.receive_buffer_empty() {
+            return Err(FpgaError::InvalidState);
+        }
+
+        // Add the segement base address to the given read offset.
+        let address = segment as u32 | offset.into();
+
+        // Write the opcode.
+        self.fpga.write(
+            WriteOp::Write,
+            Addr::TOFINO_DEBUG_PORT_BUFFER,
+            u8::from(DebugRequestOpcode::DirectRead),
+        )?;
+
+        // Write the address.
+        for b in address.as_bytes().iter() {
+            self.fpga.write(
+                WriteOp::Write,
+                Addr::TOFINO_DEBUG_PORT_BUFFER,
+                *b,
+            )?;
+        }
+
+        // Start the request.
+        self.fpga.write(
+            WriteOp::Write,
+            Addr::TOFINO_DEBUG_PORT_STATE,
+            Reg::TOFINO_DEBUG_PORT_STATE::REQUEST_IN_PROGRESS,
+        )?;
+
+        // Wait for the request to complete.
+        while self.state()?.request_in_progress() {
+            userlib::hl::sleep_for(1);
+        }
+
+        // Read the response.
+        let mut v: u32 = 0;
+        for b in v.as_bytes_mut().iter_mut() {
+            *b = self.fpga.read(Addr::TOFINO_DEBUG_PORT_BUFFER)?;
+        }
+
+        Ok(v)
+    }
+
+    pub fn write_direct(
+        &self,
+        segment: DirectBarSegment,
+        offset: impl Into<u32>,
+        value: u32,
+    ) -> Result<(), FpgaError> {
+        let state = self.state()?;
+
+        if !state.send_buffer_empty() || !state.receive_buffer_empty() {
+            return Err(FpgaError::InvalidState);
+        }
+
+        // Add the segement base address to the given read offset.
+        let address = segment as u32 | offset.into();
+
+        // Write the opcode to the queue.
+        self.fpga.write(
+            WriteOp::Write,
+            Addr::TOFINO_DEBUG_PORT_BUFFER,
+            u8::from(DebugRequestOpcode::DirectWrite),
+        )?;
+
+        // Write the address to the queue.
+        for b in address.as_bytes().iter() {
+            self.fpga.write(
+                WriteOp::Write,
+                Addr::TOFINO_DEBUG_PORT_BUFFER,
+                *b,
+            )?;
+        }
+
+        // Write the value to the queue.
+        for b in value.as_bytes().iter() {
+            self.fpga.write(
+                WriteOp::Write,
+                Addr::TOFINO_DEBUG_PORT_BUFFER,
+                *b,
+            )?;
+        }
+
+        // Start the request.
+        self.fpga.write(
+            WriteOp::Write,
+            Addr::TOFINO_DEBUG_PORT_STATE,
+            Reg::TOFINO_DEBUG_PORT_STATE::REQUEST_IN_PROGRESS,
+        )?;
+
+        // Wait for the request to complete.
+        while self.state()?.request_in_progress() {
+            userlib::hl::sleep_for(1);
+        }
+
+        Ok(())
+    }
+
+    /// Generate the SPI command Tofino needs to complete a SPI request.
+    fn spi_command(n_bytes_to_write: usize, n_bytes_to_read: usize) -> u32 {
```

Comment:
Add some assertions that `n_bytes_to_read/write` are correctly sized here

Reply:
Good idea, thanks.